src/lib.rs

```diff
@@ -102,87 +88,120 @@ pub fn demangle(mut s: &str) -> Demangle {
+                style: Some(style),
+                original: s,
+                suffix: remainder,
+            };
         }
     }
 
-    Demangle {
-        style,
+    return Demangle {
+        style: None,
         original: s,
-        suffix,
+        suffix: "",
+    };
+}
+
+#[cfg(feature = "std")]
+fn demangle_partial(s: &str) -> (Demangle, &str) {
+    if let Some((style, remainder)) = demangle_common(s) {
+        // Note: suffix is ALWAYS empty because we do not compute the
+        // LLVM compatibility (nor do we care)
+        return (
+            Demangle {
+                style: Some(style),
+                original: s,
+                suffix: "",
+            },
+            remainder,
+        );
     }
+
+    (
+        Demangle {
+            style: None,
+            original: s,
+            suffix: "",
+        },
+        s,
+    )
 }
 
 #[cfg(feature = "std")]
 fn demangle_line(
-    line: &str,
+    mut line: &str,
     output: &mut impl std::io::Write,
     include_hash: bool,
 ) -> std::io::Result<()> {
-    let mut head = 0;
-    while head < line.len() {
+    loop {
         // Move to the next potential match
-        let next_head = match (line[head..].find("_ZN"), line[head..].find("_R")) {
-            (Some(idx), None) | (None, Some(idx)) => head + idx,
-            (Some(idx1), Some(idx2)) => head + idx1.min(idx2),
+        let next_head = match (line.find("_ZN"), line.find("_R")) {
+            (Some(idx), None) | (None, Some(idx)) => idx,
+            (Some(idx1), Some(idx2)) => idx1.min(idx2),
             (None, None) => {
                 // No more matches...
                 line.len()
             }
         };
-        output.write_all(line[head..next_head].as_bytes())?;
-        head = next_head;
-        // Find the non-matching character.
-        //
-        // If we do not find a character, then until the end of the line is the
-        // thing to demangle.
-        let match_end = line[head..]
-            .find(|ch: char| !(ch == '$' || ch == '.' || ch == '_' || ch.is_ascii_alphanumeric()))
-            .map(|idx| head + idx)
-            .unwrap_or(line.len());
-
-        let mangled = &line[head..match_end];
-        head = head + mangled.len();
-        if let Ok(demangled) = try_demangle(mangled) {
+        output.write_all(line[..next_head].as_bytes())?;
+        line = &line[next_head..];
+
+        if line.is_empty() {
+            break;
+        }
+
+        let (demangled, remainder) = demangle_partial(line);
+        line = remainder;
+
+        if demangled.style.is_some() {
             if include_hash {
                 write!(output, "{}", demangled)?;
             } else {
                 write!(output, "{:#}", demangled)?;
             }
         } else {
-            output.write_all(mangled.as_bytes())?;
+            // there are maybe valid symbols inside this fake one
+            output.write_all(&line.as_bytes()[..1])?;
+            line = &line[1..];
```

Comment:
Note again a subtle difference: this can now identify and write symbols inside what appear to be other symbols.

Reply:
Wouldn't that accidentally demangle parts of C++ symbols as Rust symbols? If so that would break C++ demangling for programs that try Rust demangling before C++ demangling to handle both. Trying C++ demangling first is guaranteed to work for legacy Rust symbols, but doesn't demangle the `$LT$`,`$C$`, ... parts.